File: codes/OOP/java/Polymorphism.java
/**
 * Created by gaurav on 6/2/16.
 */
class Shape {
    protected int height, width;
    public void setHeight(int h) {
        height = h;
    }
    public void setWidth(int w) {
        width = w;
    }
    public void printArea() {
        System.out.println("Shape Unknown");
    }
    public int getWidth() {
        return width;
    }
}
class Rectangle extends Shape {
    public void printArea() {
        System.out.println(width*height);
    }
    public int getWidth() {
        return super.getWidth();
    }
}

public class Polymorphism {
    public static void main(String args[]) {

        Rectangle rectangle = new Rectangle();
        rectangle.setHeight(5);
        rectangle.setWidth(4);
        rectangle.printArea();

        Shape shape = new Rectangle();
        shape.setHeight(5);
        shape.setWidth(4);
        shape.printArea();
        System.out.println(shape.getWidth());
    }
}

File: codes/OOP/cpp/polymorphism.cpp
//A call to a member function causes a different function to be executed depending on the type of object that invokes the function.
#include<bits/stdc++.h>
using namespace std;

class Shape {
protected:
    int width, height;
public:
    Shape(int w, int h) {
        width = w;
        height = h;
    }
    virtual void print_area() {
        cout << "Shape undefined\n";
    }
};

class Rectangle: public Shape {
public:
    Rectangle(int w, int h): Shape(w, h) {};
    void print_area() {
        cout << width*height << endl;
    }
};

int main( )
{
    Shape *shape;
    Rectangle rect(5, 7);
    shape = &rect;
    shape->print_area();
   return 0;
}
